app/javascript/actions/productsActions.js

```diff
@@ -0,0 +1,181 @@
+import axios from 'axios';
+import BotRegistry from '../blockchain/BotRegistry';
+import BotCoin from '../blockchain/BotCoin';
+import { normalizeProducts } from '../helpers/JsonNormalizer';
+import * as WalletActions from './walletActions';
+
+export const ProductsActions = {
+  RESET_STATE: 'PRODUCTS_RESET_STATE',
+  APPEND: 'PRODUCTS_APPEND',
+  SET_ATTRIBUTE: "PRODUCTS_SET_ATTRIBUTE",
+  SET_PROGRESS: "PRODUCTS_SET_PROGRESS"
+}
+
+/** Resets redux state for AI products */
+export const resetTxs = () => (dispatch) => {
+  dispatch({ type: ProductsActions.RESET_STATE });
+}
+
+/** Append products in redux state
+ * @param products - normalized products (helper normalizeProducts)
+ **/
+export const appendProducts = (products) => {
+  return {
+    type: ProductsActions.APPEND,
+    products: normalizeProducts(products)
+  }
+}
+
+/** Sets in progress flag used to display in progress message or animation
+ * @param status - boolean value, true if request is in progress
+ **/
+const setInProgress = (status) => {
+  return { type: ProductsActions.SET_ATTRIBUTE, key: 'inProgress', value: status }
+}
+
+/** Sets progress status to display in progress message
+ * @param step - string with step name
+ * @param status - string with status of this step
+ **/
+const setProgressStatus = (step, status) => {
+  return { type: ProductsActions.SET_PROGRESS, step: step, status: status }
+}
+
+/** setErrors
+ * @param errors - array of errors
+ **/
+const setErrors = (errors) => {
+  return { type: ProductsActions.SET_ATTRIBUTE, key: 'errors', value: errors }
+}
+
+/** Fetch entryPrice from BotRegistry */
+export const fetchEntryPrice = () => async (dispatch) => {
+  let registry = new BotRegistry(window.app_config.bot_registry_contract);
+  let price = await registry.getEntryPrice();
+  let botCoin = new BotCoin();
+  dispatch({ type: ProductsActions.SET_ATTRIBUTE, key: 'entryPrice', value: botCoin.convertToHuman(price) });
+  if (price == 0) dispatch(setProgressStatus('approve', 'not_used'));
+}
+
+/** Upload AI product metadata to IPFS 
+ * @param values - AI product metadata as JS object
+**/
+const addMetadata2IPFS = (values) => (dispatch) => {
+  return new Promise((resolve, reject) => {
+    dispatch(setProgressStatus('load_to_ipfs', 'running'));
+    const config = { headers: { 'content-type': 'multipart/form-data' } };
+    const formData = new FormData()
+    formData.append('file', JSON.stringify(values))
+
+    axios.post('https://ipfs.infura.io:5001/api/v0/add?pin=true', formData, config)
```

Comment:
should we move 'https://ipfs.infura.io:5001/' to configs?

Reply:
I thought about this, but if we changed IPFS services, I think we should also change source code for process data from it. That's why I leave URL in code.